rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,557 @@
+
+[StreamExecutor](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_pimpl.h;l=73) is TensorFlow's main device manager, responsible for work execution and memory management. It provides a set of methods
+(such as
+[Memcpy](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=240))
+that can be customized for a particular device.
+
+We propose a C API wrapper of a subset of methods in
+[StreamExecutorInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=166?q=StreamExecutorinterface)
+as an ABI-stable way to register a custom StreamExecutor platform.
+
+# User Benefits
+
+A decoupled way to add a new device to TensorFlow.
+
+*   Simpler process: Does not have to add a new build toolchain to TensorFlow
+*   Faster time-to-solution: Does not need code review from the TensorFlow team.
+*   Lower maintenance efforts: Only C-API-related changes could break the
+    integration. Unrelated TensorFlow changes would not break the code.
+       *    The C APIs may be changed during the initial experimental phase based 
+            on developer experience and feedback. When the APIs become more mature, 
+            we will try to keep them stable (in a best-effort manner) until the new 
+            TensorFlow stack is available.
+
+# Design Proposal
+
+## StreamExecutorInterface
+
+[StreamExecutorInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=166?q=StreamExecutorinterface)
+is quite large and some of its methods are only sporadically used. Therefore, we
+plan to wrap only a subset of key StreamExecutorInterface functionality.
+
+See proposed C API below:
+
+```cpp
+#include <stddef.h>
+#include <stdint.h>
+
+#define SE_MAJOR 0
+#define SE_MINOR 0
+#define SE_REVISION 1
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef SE_Stream_st* SE_Stream;
+typedef SE_Event_st* SE_Event;
+typedef SE_Timer_st* SE_Timer;
+typedef TF_Status* (*TF_StatusCallbackFn)(void*);
+
+#ifndef TF_BOOL_DEFINED
+#define TF_BOOL unsigned char
+#endif // TF_BOOL_DEFINED
+
+#ifndef TF_OFFSET_OF_END
+#define TF_OFFSET_OF_END(TYPE, MEMBER) (offsetof(TYPE, MEMBER) + sizeof(((TYPE *)0)->MEMBER))
+#endif // TF_OFFSET_OF_END
+
+typedef struct SE_PlatformId {
+  size_t struct_size;
+  void* ext;
+  void* id;  // aka stream_executor::Platform::Id
+} SE_PlatformId;
+
+#define SE_PLATFORMID_STRUCT_SIZE TF_OFFSET_OF_END(SE_PlatformId, id)
+
+typedef struct SE_TimerFns {
+  size_t struct_size;
+  void* ext;
+  uint64_t (*nanoseconds)(SE_Timer timer);
+  uint64_t (*microseconds)(SE_Timer timer);
+} SE_TimerFns;
+
+#define SE_TIMER_FNS_STRUCT_SIZE TF_OFFSET_OF_END(SE_TimerFns, microseconds)
+
+typedef struct SE_AllocatorStats {
```

Comment:
OOC, why not just `typedef struct {`?

Reply:
Strictly speaking, the tag name doesn't have to be there since we have typedef.  It was naturally contracted from:
```
struct Foo {...};
typedef struct Foo Foo;
```
combined into one expression `typedef struct Foo  { … } Foo; `. I suppose some nice qualities about having tag names is that you can detect tag name collisions if they're there, and you can copy-paste the struct tag name without the typedef part? :man_shrugging:


(I know that you already know this but for completeness: the `typedef` is there so that in C we don’t have to declare Foo as `struct Foo foo;` )